### ISSUE TITLE
R Matrix of NA does not convert to Pandas DataFrame of nan

### ISSUE THREAD

**Issue by spluque** [rpy2/rpy2#799]
As suggested in [StackOverflow](https://stackoverflow.com/questions/67141027/how-to-ensure-a-numeric-matrix-is-retrieved-from-r/67141492#67141492), conversion of an R matrix of NA values is not handled properly within a conversion context.

**To Reproduce**

```python
import pandas as pd
import rpy2.robjects.conversion as cv
import rpy2.robjects as robjs
from rpy2.robjects.packages import importr
from rpy2.robjects import pandas2ri

r_base = importr("base")

na_mat = r_base.matrix(robjs.r.rep(robjs.r("NA"), 3), 1, 3)

with cv.localconverter(robjs.default_converter +
                       pandas2ri.converter):
    pd_df = pd.DataFrame(na_mat)
```

`pd_df` ends up as a Pandas DataFrame:

```python
            0
0 -2147483648
1 -2147483648
2 -2147483648
```

**Expected behavior**

`pd_df` should be a 1x3 DataFrame with nan values:

```python
    0   1   2
0 NaN NaN NaN
```

**Additional context**

R -> 4.0.4
Rpy2 -> 3.4.3
Python -> 3.9.2

**Comment by lgautier** [comment]
Up to relatively recently it was not possible to have NA (missing) values in pandas boolean arrays, IIRC. Behind the hood those are numpy arrays of dtype `bool`. Each item takes one byte. Zero means `False` and non-zero means `True`. 

R boolean arrays take more space (more bytes) per item and use the value `-2147483648` to mean `NA` (missing).

Pandas now has a class `BooleanArray` as a possible solution to that but it is marked as experimental at the moment. If  anyone has experience with them and can vouch for them (as in "work well in practice", and "API is now stable enough to be considered") I am happy to update the rpy2 conversion to use it.

**Comment by spluque** [comment]
I see that boolean arrays were introduced in Pandas 1.0.0 in 2020, which was a big jump (it bit me as a user in few places).  I understand why the conversion will have to be from R boolean to Pandas boolean, so it seems it will be up to the user to convert to numeric nan regardless of the solution, correct?

**Comment by lgautier** [comment]
For now it is left to the user because I am not familiar enough with `BooleanArray` to understand what introducing them to the default pandas converter in rpy2 would mean in terms of support. The current converter is imperfect because it converts R booleans to signed integers, with NAs taking a special negative value, but it is otherwise stable. I don't know what I'd get into with an experimental class.

In the meantime the conversion system is meant to be modular. You could add a custom conversion rule with `BooleanArray` and see how it goes.

**Comment by lgautier** [comment]
`pandas` itself is kind of surprising or inconsistent. `None` is used to mean `NA`... some of times.

With `bool`, `None` is `False`:

```python
from pandas.core.series import Series
s = Series([True, False, None], index=['a', 'b', 'c'], dtype='bool')
repr(s)
```
```
a     True
b    False
c    False
dtype: bool
```

With `int`, it is an exception:

```python
 s = Series([True, False, None], index=['a', 'b', 'c'], dtype='int')
```
```
TypeError: int() argument must be a string, a bytes-like object or a number, not 'NoneType'
```

With `float` it is `nan`:
 
```python
s = Series([True, False, None], index=['a', 'b', 'c'], dtype='float')
repr(s)
```

```
a    1.0
b    0.0
c    NaN
dtype: float64
```

**Comment by lgautier** [comment]
I just had a look and the pandas converter did not use an additional dispatch mechanism added recently to the conversion system.
If going down the path of having custom converter, check 00f411f928a4d6d79fa52c07f74e084bd8f6afee.

**Comment by spluque** [comment]
> For now it is left to the user because I am not familiar enough with `BooleanArray` to understand what introducing them to the default pandas converter in rpy2 would mean in terms of support. The current converter is imperfect because it converts R booleans to signed integers, with NAs taking a special negative value, but it is otherwise stable. I don't know what I'd get into with an experimental class.
> 
> In the meantime the conversion system is meant to be modular. You could add a custom conversion rule with `BooleanArray` and see how it goes.

Yes, it looks like this is still flagged as experimental in the current stable Pandas (1.2.4), so this seems wise.  In fact, reading the NA help page in R, NAs can be coerced to many other forms, so one really has to be super careful on R's end anyway.  Using the `na_mat` object below does yield the desired output:

```python
na_mat = r_base.as_numeric(
    r_base.matrix(robjs.r.rep(robjs.r("NA"), 3), 1, 3))
```

**Comment by alonme-loops** [comment]
@lgautier  - Is there any reason not to use `np.NAN` / `pd.NA` instead of `2147483648` ?

**Comment by alonme** [comment]
@lgautier Is there any reason not to use `np.NaN` or `pd.NA` instead of `2147483648` ?

**Comment by lgautier** [comment]
@alonme : it is actually `-2147483648`. This is the value that R has elected to represent missingness (NA) for integer arrays. There is otherwise no IEEE standard to represent this, or a special value like `nan` for floats that can be used to that effect. 

R arrays or type float or int are essentially wrapper around C arrays of the same type. When considering int arrays, the use of `np.NaN` is not possible as this special is a float (and should it be possible to be casted down to the right type for R integer arrays it has no special meaning among integers). You can do that in numpy either:
```python
>>> import numpy
>>> a = numpy.ndarray((1,2,3), dtype=numpy.int32)
>>> a[0] = numpy.NaN
ValueError: cannot convert float NaN to integer
>>> a = numpy.ndarray((1,2,3), dtype=numpy.int64)
>>> a[0] = numpy.NaN
ValueError: cannot convert float NaN to integer
```

`pd.NA` is a Python object. Using it will not work with int arrays backed by a C array of matching type. 

```python
>>> a[0] = pandas.NA
TypeError: int() argument must be a string, a bytes-like object or a number, not 'NAType'
```

**Comment by alonme** [comment]
@lgautier  - thanks, missing the minus was a typo 😄 

As i look at it, i am using an `rpy2` to pandas converter, so i am expecting the result to behave like pandas does.

for example - you can add an NaN value to an int column - with the following behavior.

```
In [35]: import pandas as pd

In [36]: import numpy as np

In [37]: df1 = pd.DataFrame({'a': [1,2,3]})

In [38]: df1['a']
Out[38]:
0    1
1    2
2    3
Name: a, dtype: int64

In [39]: df1['a'][0] = np.NaN

In [40]: df1['a']
Out[40]:
0    NaN
1    2.0
2    3.0
Name: a, dtype: float64
```

Also - given a column which only contains pd.NA or np.NaN, there are also defined behaviors (which are different than what happens with a `NA` from rpy2)



```
In [43]: import pandas as pd
    ...: import numpy as np
    ...:
    ...:
    ...: import rpy2.robjects as ro
    ...: from rpy2.robjects import pandas2ri
    ...: from rpy2.robjects.conversion import localconverter
    ...:
    ...: r_df = ro.r(
    ...:     """
    ...:     res <- data.frame(a = NA)
    ...:     """
    ...: )
    ...:
    ...:
    ...: with localconverter(ro.default_converter + pandas2ri.converter):
    ...:     py_from_r_df = ro.conversion.rpy2py(r_df)
    ...:
    ...: display(py_from_r_df)
    ...: print('types')
    ...: display(py_from_r_df.dtypes)
            a
1 -2147483648
types
a    int32
dtype: object

In [44]: df_NA = pd.DataFrame({'a': [pd.NA]})
    ...: df_NaN = pd.DataFrame({'a': [np.NaN]})
    ...:
    ...:
    ...: display(df_NA)
    ...: print('types')
    ...: display(df_NA.dtypes)
    ...: display(df_NaN)
    ...: print('types')
    ...: display(df_NaN.dtypes)
      a
0  <NA>
types
a    object
dtype: object
    a
0 NaN
types
a    float64
dtype: object
```
